Move promotion rule application to C front-end [blocks: #3725, #3800]

tautschnig opened the pull request:
The front-ends handle type conversion when doing arithmetic over enum types. The
case of an enum tag is not expected in goto-program conversion. That, however,
was only the case for some operators. Made the C front-end properly handle all
assignment operators.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Move promotion rule application to C front-end

The front-ends handle type conversion when doing arithmetic over enum types. The
case of an enum tag is not expected in goto-program conversion. That, however,
was only the case for some operators. Made the C front-end properly handle all
assignment operators.

While at it, also support decrement operators over complex numbers. We already
did so for increment, decrement just requires the same special case.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 965d458bebe8f43b8293d6f4417be20f1d67c4e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97133012

tautschnig commented:
Marked work-in-progress: we should review all special-casing of `ID_c_enum_tag` (and indeed also `ID_c_bool`) in `goto-programs`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: aa1b446957bc70305bff841a5bd9c6c91b011b06).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99429688

tautschnig committed:
Move LHS type casts to RHS in assign_* instructions

Doing so during goto-program conversion will ensure we can generate appropriate
conversion checks.

Fixes: #4208

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9f92ab9271ae6c1d2a1bc0c6930a2ee59354e619).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101215749

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3d01c87de20836aaa298e6ee090718d856e6023c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101250611

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e254a6f49c05dd0937176c0b10bd3e8df47b57bd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101349909

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1042845d8721cfe3fb7751c12582cc444a6e95aa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102431450

kroening commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1846,16 +1846,19 @@ void c_typecheck_baset::typecheck_expr_side_effect(side_effect_exprt &expr)
+
+      // increment/decrement on underlying type
+      expr.op0() = typecast_exprt(expr.op0(), enum_type.subtype());
+      expr.type() = enum_type.subtype();
```
I would make that an implicit typecast.

tautschnig replied:
I have added `implicit_typecast_arithmetic(expr);` - is that what you expected?

kroening replied:
No, that's arithmetic promotion; I am not actually sure whether these are meant to happen to the underlying type of an enum.
What I meant is the `ID_C_implicit_typecast` annotation, which is used to distinguish an explicit cast from one added by the frontend.

tautschnig replied:
I can't seem to find `ID_C_implicit_typecast` or indeed any other annotation that the `implicit_typecast` code would be adding?!

kroening commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -99,33 +102,19 @@ void goto_convertt::remove_assignment(
+
+    exprt new_lhs = skip_typecast(expr.op0());
+    rhs = typecast_exprt::conditional_cast(rhs, new_lhs.type());
+    rhs.add_source_location() = expr.source_location();
```
Same here.

tautschnig replied:
This is outside the C front-end, I don't think I can use `implicit_typecast(_arithmetic)` here?!

kroening replied:
Yes, you may have to add `ID_C_implicit_typecast` directly.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 25e854f2a865d8b54d0114309dde4540f620dd1a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102514059

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20430258d1dbfc4cb2301d178cc91b33ef79948b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102521796
